BHoM_Engine/Compute/DefiningString.cs

```diff
@@ -0,0 +1,149 @@
+/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2020, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.Engine.Base.Objects;
+using BH.oM.Base;
+using BH.oM.Reflection.Attributes;
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Data;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+
+namespace BH.Engine.Base
+{
+    public static partial class Compute
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Generates a string representing the whole structure of the object with its assigned values.")]
+        [Input("obj", "Objects the string should be calculated for.")]
+        [Input("propertyNameExceptions", "(Optional) e.g. `Fragments`. If you want to exclude a property that many objects have.")]
+        [Input("propertyFullNameExceptions", "(Optional) e.g. `BH.oM.Structure.Elements.Bar.Fragments`. If you want to exclude a specific property of an object.")]
+        [Input("namespaceExceptions", "(Optional) e.g. `BH.oM.Structure`. Any corresponding namespace is ignored.")]
+        [Input("typeExceptions", "(Optional) e.g. `typeof(Guid)`. Any corresponding type is ignored.")]
+        [Input("maxNesting", "(Optional) e.g. `100`. If any property is nested into the object over that level, it is ignored.")]
+        public static string DefiningString(
```

Comment:
Reason why I kept it public: try it by itself! 😄 
It returns a nice "serialised" view of the object. Since I was there, I thought it could be useful. I'd like to keep it public.

Reply:
I though that would be the reason to have it public 😋 
That doesn't change my opinion though. The BH.Base namespace is very small and tidy as it exposes the fundamentals around BHoM objects. Exposing a method there that has no meaning and direct use for anyone but you is going against that in my opinion.

Reply:
> Exposing a method there that has no meaning and direct use for anyone but you is going against that in my opinion.

that is what I am contending against 😄 I still think it can be useful in cases you want to drop a quick view of the object without having to serialise it, e.g. in the immediate window of VS. Let's hear others opinions @al-fisher and then I'll be happy to privatise it if deemed better.

Reply:
Looking at this - agreed @alelom - can see a use case for simple string view of an instance of any object, including all property values.
Think the method name and namespace might need review to make intuitive in that sense.
i.e. ToString inline with ToJson ?
So in current form with all arguments think perhaps worth keeping private as is in direct support of the Hash.
Can raise a separate issue to close a ToString (or similar) method as you describe @alelom as do think that will be useful in a few places